docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,282 @@
+    wing.cloud/login/callback-->>wing.cloud/login/callback: Exchange tokens and encode them in a cookie for https://wing.cloud
+    wing.cloud/login/callback-)wing.cloud: Redirect (now with the authenticated cookie)
+```
+
+> **Important Note:** Since we don't ask for information during the sign up process, both the sign up and sign in flows are the same.
+
+#### Process: Authentication and Authorization
+
+The website https://wing.cloud will use a secure HTTP-only cookie that safely encodes a JWT containing the Wing Cloud user ID. We could use the Github user ID, but we may want to support other authentication providers in the future.
+
+Every possible interaction with the Wing Cloud will be through an API. Any request to the API will be authenticated using this cookie. The API will then use the user ID to check if the user has the necessary permissions to perform the action.
+
+Since we want to go minimal, we can use the GitHub API and the user permissions to the repository as the authorization mechanism. Beware the GitHub API has a [rate limit](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/rate-limits-for-github-apps#installation-access-tokens-on-githubcom), so we may want to cache the permissions for a while if we foresee hitting these limits. Also, see the [List collaborators endpoint](https://docs.github.com/en/free-pro-team@latest/rest/collaborators/collaborators?apiVersion=2022-11-28#list-repository-collaborators) for more information on authorization.
+
+#### Process: Processing GitHub Events
+
+A Cloud.Function will listen to GitHub events and will process them. Based on the event, the function will create, update or delete preview environments on Flyio.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    flyio{Flyio}
+    gh -- Integration --> ghapp
+    ghapp -- Events --> fn
+    fn -- Create Preview Environment --> flyio
+    fn -- Update Preview Environment --> flyio
+    fn -- Delete Preview Environment --> flyio
+```
+
+Depending on whether GitHub retries events or not, we may want to include a FIFO queue in the middle.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    queue[(Events Queue)]
+    fn2[Enqueue Events Fn]
+    ghapp -- Events --> fn2
+    fn2 --> queue
+    queue --> fn
+    gh -- Integration --> ghapp
+    fn --> rest[...]
+```
+
+The probot npm package will take care of ensuring the events received are legitimate.
+
+#### Process: Creating Preview Environments
+
+When a GitHub event `pull_request.opened` is received, we will look for `*.main.w` files and create a preview environment for each of them. We will use the Flyio API to create a new project(s) and deploy the code to it (the Flyio specifics may change).
+
+During this process, we will also create new entries in the Cloud.Table to store the Flyio project ID, the GitHub PR ID, and the entry point.
```

Comment:
can you please explain why do we need to save this information?
(we discussed about reducing the gh api requests for minimum - this was one reason for saving the PR and the VM ID, for knowing which PR to update if the VM was stale for X amount of time)

Reply:
I rephrased everything a bit, but those reasons are correct. Another reason is that we have to resolve a domain such as `<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app` or `<gh-repository>-<nanoid>.wingcloud.app` to a Flyio domain.